crates/dns/src/kube.rs

```diff
@@ -0,0 +1,251 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use trust_dns_client::rr::LowerName;
+use trust_dns_proto::error::{ProtoError, ProtoResult};
+use trust_dns_proto::rr::Name;
+
+// TODO(nmittler): This should be re-written to support separate name tables for in-namespace and in-domain.
+
+/// Information about the client workload.
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct ClientInfo {
+    /// The namespace of the client.
+    pub namespace: String,
+
+    /// The domain of the client. This will typically be `svc.cluster.local`.
+    pub domain: Name,
+}
+
+/// Information about the target service.
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct ServiceInfo {
+    /// The FQDN for the service. Typically of the form `<name>.<namespace>.svc.cluster.local`.
+    pub fqdn: Name,
+
+    /// The name of the service.
+    pub name: String,
+
+    /// The namespace of the service.
+    pub namespace: String,
+}
+
+/// Generates the set of valid alternative hosts for the service FQDN
+/// (e.g. `s1.ns1.svc.cluster.local`), based on the location of the client.
```

Comment:
```suggestion
/// (e.g. `s1.ns1.svc.cluster.local`), based on the location of the client.
```

The example is not a FQDN in the context of DNS/the DNS library we use. FQDN means has a trailing `.`